pinot-core/src/main/java/org/apache/pinot/core/segment/processing/serde/GenericRowSerializer.java

```diff
@@ -0,0 +1,228 @@
+            byte[] stringBytes = StringUtils.encodeUtf8((String) value);
+            numBytes += Integer.BYTES + stringBytes.length;
+            _stringBytes[i] = stringBytes;
+            break;
+          case BYTES:
+            numBytes += Integer.BYTES + ((byte[]) value).length;
+            break;
+          default:
+            throw new IllegalStateException("Unsupported SV stored type: " + _storedTypes[i]);
+        }
+      } else {
+        Object[] multiValue = (Object[]) value;
+        int numValues = multiValue.length;
+        numBytes += Integer.BYTES; // Number of values
+
+        switch (_storedTypes[i]) {
+          case INT:
+            numBytes += Integer.BYTES * numValues;
+            break;
+          case LONG:
+            numBytes += Long.BYTES * numValues;
+            break;
+          case FLOAT:
+            numBytes += Float.BYTES * numValues;
+            break;
+          case DOUBLE:
+            numBytes += Double.BYTES * numValues;
+            break;
+          case STRING:
+            numBytes += Integer.BYTES * numValues;
+            byte[][] stringBytesArray = new byte[numValues][];
+            for (int j = 0; j < numValues; j++) {
+              byte[] stringBytes = StringUtils.encodeUtf8((String) multiValue[j]);
+              numBytes += stringBytes.length;
+              stringBytesArray[j] = stringBytes;
+            }
+            _stringBytes[i] = stringBytesArray;
+            break;
+          default:
+            throw new IllegalStateException("Unsupported MV stored type: " + _storedTypes[i]);
+        }
+      }
+    }
+
+    // Serialize null value fields if enabled
+    if (_fieldIndexMap != null) {
+      Set<String> nullValueFields = row.getNullValueFields();
+      numBytes += Integer.BYTES * (1 + nullValueFields.size());
+    }
+
+    byte[] serializedBytes = new byte[numBytes];
+    ByteBuffer byteBuffer = ByteBuffer.wrap(serializedBytes).order(PinotDataBuffer.NATIVE_ORDER);
+
+    // Second pass: serialize the values
+    for (int i = 0; i < _numFields; i++) {
+      Object value = row.getValue(_fieldNames[i]);
+
+      if (_isSingleValueFields[i]) {
+        switch (_storedTypes[i]) {
+          case INT:
+            byteBuffer.putInt((int) value);
+            break;
+          case LONG:
+            byteBuffer.putLong((long) value);
+            break;
+          case FLOAT:
+            byteBuffer.putFloat((float) value);
+            break;
+          case DOUBLE:
+            byteBuffer.putDouble((double) value);
+            break;
+          case STRING:
+            byte[] stringBytes = (byte[]) _stringBytes[i];
+            byteBuffer.putInt(stringBytes.length);
+            byteBuffer.put(stringBytes);
+            break;
+          case BYTES:
+            byte[] bytes = (byte[]) value;
+            byteBuffer.putInt(bytes.length);
+            byteBuffer.put(bytes);
+            break;
+          default:
+            throw new IllegalStateException();
```

Comment:
nit: even though this default block will never be reached, add message to the exception? for code readability and also to catch any mistakes made by future development